### ISSUE TITLE
[torch.compile] `fuse_attention` returns inconsistent value for the model

### ISSUE THREAD

**Issue by YangChenyuan** [pytorch/pytorch#121174]
### 🐛 Describe the bug

With `torch.compile`, `fuse_attention` returns inconsistent value for the model

```py
import torch
import math

torch.manual_seed(0)

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.q_conv = torch.nn.Conv2d(4, 4, 1)
        self.k_conv = torch.nn.Conv2d(4, 4, 1)
        self.v_conv = torch.nn.Conv2d(4, 4, 1)

    def forward(self, x):
        q = self.q_conv(x)
        k = self.k_conv(x)
        v = self.v_conv(x)
        q = q.permute(0, 2, 1, 3)
        k = k.permute(0, 2, 1, 3)
        v = v.permute(0, 2, 1, 3)
        div = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(q.size(-1))
        attn_weight = torch.nn.functional.softmax(div, dim=-1)
        output = torch.matmul(attn_weight, v)
        return output

func = Model()

x = torch.randn(1, 4, 2, 2)

func = Model()

with torch.no_grad():
    print(func(x.clone()))
    # tensor([[[[ 0.1035, -0.2616],
    #         [ 0.0461, -0.3923],
    #         [ 0.0852, -0.4379],
    #         [ 0.0538, -0.5618]],
    #         [[-0.3332, -0.2985],
    #         [-0.1777, -0.2978],
    #         [-0.4235, -0.3401],
    #         [-0.4361, -0.3477]]]])

    func1 = torch.compile(func)
    print(func1(x.clone()))
    # tensor([[[[ 0.1624,  0.2315],
    #         [ 0.0282,  0.0822],
    #         [ 0.2613,  0.0440],
    #         [-0.0161, -0.0560]],

    #         [[-0.3225, -0.2300],
    #         [-0.2753, -0.3049],
    #         [-0.3932, -0.3184],
    #         [-0.3870, -0.2880]]]])

    print(torch.equal(func(x.clone()), func1(x.clone()))) # False
    print(torch.equal(func.k_conv.weight, func1.k_conv.weight)) # True
    print(torch.equal(func.q_conv.weight, func1.q_conv.weight)) # True
    print(torch.equal(func.v_conv.weight, func1.v_conv.weight)) # True
```

If I disable the `fuse_attention`, the compiled model will return the same value without optimization.

### Versions

PyTorch version: 2.3.0.dev20240301+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.0 (default, Nov 15 2020, 14:28:56) [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-6.5.0-21-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 24
On-line CPU(s) list: 0-23
Vendor ID: GenuineIntel
Model name: 12th Gen Intel(R) Core(TM) i9-12900K
CPU family: 6
Model: 151
Thread(s) per core: 2
Core(s) per socket: 16
Socket(s): 1
Stepping: 2
CPU max MHz: 5200.0000
CPU min MHz: 800.0000
BogoMIPS: 6374.40
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization: VT-x
L1d cache: 640 KiB (16 instances)
L1i cache: 768 KiB (16 instances)
L2 cache: 14 MiB (10 instances)
L3 cache: 30 MiB (1 instance)
NUMA node(s): 1
NUMA node0 CPU(s): 0-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-triton==3.0.0+901819d2b6
[pip3] torch==2.3.0.dev20240301+cu118
[pip3] torchaudio==2.2.0.dev20240301+cu118
[pip3] torchvision==0.18.0.dev20240301+cu118
[conda] numpy 1.26.4 pypi_0 pypi
[conda] pytorch-triton 3.0.0+901819d2b6 pypi_0 pypi
[conda] torch 2.3.0.dev20240301+cu118 pypi_0 pypi
[conda] torchaudio 2.2.0.dev20240301+cu118 pypi_0 pypi
[conda] torchvision 0.18.0.dev20240301+cu118 pypi_0 pypi

cc @ezyang @gchanan @zou3519 @kadeng @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

**Comment by eellison** [comment]
Could we add the bug label and or silent incorrectness labels please when triaging ? 

The repro passes when we set forcibly skip the cpu attention kernel so it looks like a bug in the kernel implementation. cc @Valentine233 


```
diff --git a/torch/_inductor/pattern_matcher.py b/torch/_inductor/pattern_matcher.py
index c48251fc352..8cc57e905c0 100644
--- a/torch/_inductor/pattern_matcher.py
+++ b/torch/_inductor/pattern_matcher.py
@@ -1052,10 +1052,12 @@ def register_replacement(
                 scalar_workaround=scalar_workaround,
             )
             specific_pattern_match = specific_pattern.match(match.output_nodes()[0])  # type: ignore[arg-type]
-            if specific_pattern_match and extra_check(specific_pattern_match):
-                # trace the pattern using the shapes from the user program
-                match.replacement_graph = trace_fn(replace_fn, args)  # type: ignore[assignment]
-                return True
+            from torch.nn.attention import sdpa_kernel, SDPBackend
+            with sdpa_kernel(backends=[SDPBackend.MATH]):
+                if specific_pattern_match and extra_check(specific_pattern_match):
+                    # trace the pattern using the shapes from the user program
+                    match.replacement_graph = trace_fn(replace_fn, args)  # type: ignore[assignment]
+                    return True
             return False
 
     def normalize_args(**kwargs):
```

**Comment by Valentine233** [comment]
The issue is caused by [the transformation of channel_last layout in conv](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/conv.py#L381), making the last dim not contiguous. In fact, when [checking for flash attention feasibility](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/sdp_utils_cpp.cpp#L35), we have ensured that the last dim stride of inputs is 1. However, the `channel_last` layout transformation happens after the checking.

I suppose that CUDA would also have the issue. Could you please confirm for this? @eellison Thanks!

**Comment by eellison** [comment]
```
import torch
import math

torch.manual_seed(0)

class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()
        self.q_conv = torch.nn.Conv2d(4, 4, 1)
        self.k_conv = torch.nn.Conv2d(4, 4, 1)
        self.v_conv = torch.nn.Conv2d(4, 4, 1)

    def forward(self, x):
        q = self.q_conv(x)
        k = self.k_conv(x)
        v = self.v_conv(x)
        q = q.permute(0, 2, 1, 3)
        k = k.permute(0, 2, 1, 3)
        v = v.permute(0, 2, 1, 3)
        div = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(q.size(-1))
        attn_weight = torch.nn.functional.softmax(div, dim=-1)
        output = torch.matmul(attn_weight, v)
        return output

func = Model()

x = torch.randn(1, 4, 2, 2)

func = Model()

with torch.no_grad():
    # tensor([[[[ 0.1035, -0.2616],
    #         [ 0.0461, -0.3923],
    #         [ 0.0852, -0.4379],
    #         [ 0.0538, -0.5618]],
    #         [[-0.3332, -0.2985],
    #         [-0.1777, -0.2978],
    #         [-0.4235, -0.3401],
    #         [-0.4361, -0.3477]]]])

    func1 = torch.compile(func)
    # tensor([[[[ 0.1624,  0.2315],
    #         [ 0.0282,  0.0822],
    #         [ 0.2613,  0.0440],
    #         [-0.0161, -0.0560]],

    #         [[-0.3225, -0.2300],
    #         [-0.2753, -0.3049],
    #         [-0.3932, -0.3184],
    #         [-0.3870, -0.2880]]]])


    torch.testing.assert_close(func(x.clone()), func1(x.clone()))
    torch.testing.assert_close(func.k_conv.weight, func1.k_conv.weight) # True
    torch.testing.assert_close(func.q_conv.weight, func1.q_conv.weight) # True
    torch.testing.assert_close(func.v_conv.weight, func1.v_conv.weight) # True
```

This passes on cuda but not on cpu.

**Comment by eellison** [comment]
@Valentine233 you might need to do add the cpu sdpa op to this https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2199

**Comment by Valentine233** [comment]
> @Valentine233 you might need to do add the cpu sdpa op to this https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2199

Thanks, @eellison. This could fix the issue. Would request you to review later.

**Comment by ezyang** [comment]
Raising priority due to being an accuracy problem on an important operator

**Comment by drubinstein** [comment]
Hi. I believe I may be running into a similar issue. How did you disable fuse_attention?
